crates/turborepo-lib/src/engine/mod.rs

```diff
@@ -0,0 +1,112 @@
+mod builder;
+
+use std::{
+    collections::{HashMap, HashSet},
+    rc::Rc,
+};
+
+pub use builder::EngineBuilder;
+use petgraph::Graph;
+
+use crate::{run::task_id::TaskId, task_graph::TaskDefinition};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum TaskNode {
+    Root,
+    Task(TaskId<'static>),
+}
+
+impl From<TaskId<'static>> for TaskNode {
+    fn from(value: TaskId<'static>) -> Self {
+        Self::Task(value)
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct Building;
+#[derive(Debug, Default)]
+pub struct Built;
```

Comment:
Build state is slightly unnecessary, but I do want to discourage mutable use of the engine after construction to simplify when we execute tasks.